### PR TITLE
K8SPSMDB-265: Configure backup node priorities for PBM

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPSMDB-265](https://jira.percona.com/browse/K8SPSMDB-265) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

PBM 1.5.0 introduced a new feature to influence the backup node
selection. It selects the node with highest priority in the config.

As part of cross cluster replication changes, we need to make sure we
start backups on nodes that are part of the operator's own cluster. With
this commit, we're assigning lower priority (0.5) to external nodes so
they won't be selected by PBM.

According to PBM's implementation, if you pass a subset of the nodes,
all remaining nodes will be automatically considered as priority 1
including the primary. It's not ideal, since we don't want to take
backups on primary (which why the PBM itself assigns priority 0.5 to the
primary node if user doesn't specify anything) and we don't want to
change the backup behavior for the users. That's also the reason of the
changes about mongo client functions. It was only possible to use them
from the main controller, but I needed to initialize a client in both
backup and restore controllers.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | failed |
| scheduled-backup | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| operator-self-healing-chaos | failed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/8338cda3a250211027aa7accee44d72730429f68
image: `perconalab/percona-server-mongodb-operator:PR-750-8338cda3`

**Review by delgod — APPROVED** [review]

File: pkg/psmdb/backup/pbm.go [hunk 1]
```diff
@@ -104,34 +141,31 @@ func (b *PBM) SetConfig(stg api.BackupStorageSpec, pitr api.PITRSpec) error {
...
-		if err != nil {
-			return errors.Wrap(err, "write config")
-		}
 	case api.BackupStorageFilesystem:
 		return errors.New("filesystem backup storage not supported yet, skipping storage name")
 	default:
 		return errors.New("unsupported backup storage type")
 	}
 
+	conf.Backup.Priority = priority
```
Comment (qJkee) [inline pkg/psmdb/backup/pbm.go:163]: you can set it on line 129 when you are init the config

File: pkg/psmdb/tls/tls.go [hunk 2]
```diff
@@ -100,3 +107,34 @@ func Issue(hosts []string) (caCert []byte, tlsCert []byte, tlsKey []byte, err er
...
+	cert, err := tls.X509KeyPair(certSecret.Data["tls.crt"], certSecret.Data["tls.key"])
+	if err != nil {
+		return tls.Config{}, errors.Wrap(err, "load keypair")
+	}
+	clientCerts = append(clientCerts, cert)
+
+	return tls.Config{
+		InsecureSkipVerify: true,
+		RootCAs:            pool,
+		Certificates:       clientCerts,
```
Comment (qJkee) [inline pkg/psmdb/tls/tls.go:138]: `Certificates: []tls.Certificate{cert}`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/8338cda3a250211027aa7accee44d72730429f68
image: `perconalab/percona-server-mongodb-operator:PR-750-8338cda3`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| operator-self-healing | passed |
| upgrade | passed |
| smart-update | passed |
| operator-self-healing-chaos | passed |
| upgrade-sharded | passed |
| version-service | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/d5ef96cb3544f66174c49b8e2060310b71247fde
image: `perconalab/percona-server-mongodb-operator:PR-750-d5ef96cb`

File: pkg/controller/perconaservermongodbbackup/backup.go [hunk 3]
```diff
@@ -84,10 +84,31 @@ func (b *Backup) Start(cr *api.PerconaServerMongoDBBackup) (api.PerconaServerMon
 func (b *Backup) Status(cr *api.PerconaServerMongoDBBackup) (api.PerconaServerMongoDBBackupStatus, error) {
 	status := cr.Status
 
-	meta, err := b.pbm.C.GetBackupMeta(cr.Status.PBMname)
-	if err != nil {
-		return status, errors.Wrap(err, "get pbm backup meta")
+	var meta *pbm.BackupMeta
+	var retries uint64 = 0
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/controller/perconaservermongodbbackup/backup.go]: I changed the code as you want in https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/dc51cc347ec473eab8dcf600b9ea825a6abbfca3 but deleted this part altogether in https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/1a2e6be139b92962d503ce8abb63da468798c697

File: pkg/controller/perconaservermongodbbackup/backup.go [hunk 4]
```diff
@@ -84,10 +84,31 @@ func (b *Backup) Start(cr *api.PerconaServerMongoDBBackup) (api.PerconaServerMon
...
+			break
+		}
+
+		if errors.Is(err, pbm.ErrNotFound) {
+			log.Info("Waiting for backup metadata", "PBM name", cr.Status.PBMname, "backup", cr.Name)
+		} else {
+			return status, errors.Wrap(err, "get pbm backup meta")
+		}
+
+		if retries > 59 {
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/controller/perconaservermongodbbackup/backup.go]: same here

File: pkg/controller/perconaservermongodbbackup/backup.go [hunk 5]
```diff
@@ -84,10 +84,31 @@ func (b *Backup) Start(cr *api.PerconaServerMongoDBBackup) (api.PerconaServerMon
 func (b *Backup) Status(cr *api.PerconaServerMongoDBBackup) (api.PerconaServerMongoDBBackupStatus, error) {
 	status := cr.Status
 
-	meta, err := b.pbm.C.GetBackupMeta(cr.Status.PBMname)
-	if err != nil {
-		return status, errors.Wrap(err, "get pbm backup meta")
+	var meta *pbm.BackupMeta
+	var retries uint64 = 0
+	ticker := time.NewTicker(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/controller/perconaservermongodbbackup/backup.go]: same here

File: pkg/psmdb/tls/tls.go (hunk 2 above)
Comment (egegunes) [reply pkg/psmdb/tls/tls.go]: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/dc51cc347ec473eab8dcf600b9ea825a6abbfca3

File: pkg/psmdb/backup/pbm.go (hunk 1 above)
Comment (egegunes) [reply pkg/psmdb/backup/pbm.go]: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/dc51cc347ec473eab8dcf600b9ea825a6abbfca3

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| operator-self-healing | passed |
| smart-update | passed |
| upgrade | passed |
| operator-self-healing-chaos | passed |
| version-service | passed |
| upgrade-sharded | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| rs-shard-migration | failed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/1a2e6be139b92962d503ce8abb63da468798c697
image: `perconalab/percona-server-mongodb-operator:PR-750-1a2e6be1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| self-healing-chaos | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| operator-self-healing | passed |
| upgrade | passed |
| liveness | passed |
| upgrade-sharded | passed |
| operator-self-healing-chaos | passed |
| smart-update | passed |
| version-service | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/750/commits/b572747a6802c89e9a2f3accdef99dddad4729b2
image: `perconalab/percona-server-mongodb-operator:PR-750-b572747a`

**Pull request merged**